### PR TITLE
50/50

### PR DESCRIPTION
**Pull request by N8BWert**
## Description
The ball handler tactic randomly assigns a robot to either pass or shoot (equal probability).

### Test Case 1
1. Run Keepaway
2. Observe the robots passing and shooting
3. Profit

Expected result: The robots should pass and shoot

### CODE REVIEW

**Review by sseye6 — COMMENTED** [review]
In most cases it works; however, there are some cases in which there is a robot from our team blocking the pass in between two other robots. Then, none of the robots can do anything else. probably should be addressed.

File: rj_gameplay/rj_gameplay/tactic/ball_handler_tactic.py [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+    RECEIVE = auto()
+    INIT_SHOOT = auto()
+    INIT_SHOOT_TICK = auto()
+    SHOOT = auto()
+    DONE = auto()
+
+AGGRESSIVENESS = 1.01
+
+# the angle cutoff for people intersepting passes or shots
+CUTOFF_ANGLE = math.cos(math.pi / 10)
```
Comment (HussainGynai) [inline rj_gameplay/rj_gameplay/tactic/ball_handler_tactic.py]: This isn't an angle, its the cosine of an angle

File: rj_gameplay/rj_gameplay/tactic/ball_handler_tactic.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+            self._needs_assign = True
+            self._state = State.INIT_TICK
+
+        elif self._state == State.INIT_TICK:
+            if len(self.assigned_roles) == 1:
+                self.init_roles(world_state)
+                self._state = State.BALL_SECURE
+
+        elif self._state == State.BALL_SECURE:
+            # TODO: these lines are a little ugly, any fix?
```
Comment (HussainGynai) [inline rj_gameplay/rj_gameplay/tactic/ball_handler_tactic.py:107]: What does this mean?

File: rj_gameplay/rj_gameplay/tactic/ball_handler_tactic.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+            ball_move_role = self.assigned_roles[0]
+            intent = ball_move_role.tick(world_state)
+
+            role_intents = [(ball_move_role.robot.id, intent)]
+
+            # Calculate best shot and find nearest robot to the path and compare to threshold
+            best_shot = striker.StrikerRole()._find_target_point(world_state = world_state, kick_speed = 4.0)
+
+            if self.check_kick(world_state, best_shot, ball_move_role.robot_id):
+                print("SHOOOOOOOOOOOOOTTTTTTING!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
```
Comment (HussainGynai) [inline rj_gameplay/rj_gameplay/tactic/ball_handler_tactic.py]: Get rid of debug print

File: rj_gameplay/rj_gameplay/tactic/ball_handler_tactic.py [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+                #pass ball to most forward person that can be passed to
+                pass_targets = self.get_pass_order(world_state, ball_move_role.robot_id)
+                for robot in pass_targets:
+                    if self.check_kick(world_state, robot.pose[0:2], ball_move_role.robot_id):
+                        self._state = State.INIT_PASS
+                        self.target_pass_robot = robot
+                        break
+        
+        elif self._state == State.INIT_SHOOT:
+            print('init shoot')
```
Comment (HussainGynai) [inline rj_gameplay/rj_gameplay/tactic/ball_handler_tactic.py]: Delete debug print

File: rj_gameplay/rj_gameplay/tactic/ball_handler_tactic.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+                self._state = State.POSSESSING
+
+        elif self._state == State.POSSESSING:
+            ball_move_role = self.assigned_roles[0]
+            intent = ball_move_role.tick(world_state)
+
+            role_intents = [(ball_move_role.robot.id, intent)]
+
+            # Calculate best shot and find nearest robot to the path and compare to threshold
+            best_shot = striker.StrikerRole()._find_target_point(world_state = world_state, kick_speed = 4.0)
```
Comment (HussainGynai) [inline rj_gameplay/rj_gameplay/tactic/ball_handler_tactic.py]: This line errors and causes sim to crash.

File: rj_gameplay/rj_gameplay/tactic/ball_handler_tactic.py [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+
+            elif ball_move_role.is_done(world_state):
+                self._state = State.INIT_SHOOT
+
+            else:
+                #pass ball to most forward person that can be passed to
+                pass_targets = self.get_pass_order(world_state, ball_move_role.robot_id)
+                for robot in pass_targets:
+                    if self.check_kick(world_state, robot.pose[0:2], ball_move_role.robot_id):
+                        self._state = State.INIT_PASS
```
Comment (HussainGynai) [inline rj_gameplay/rj_gameplay/tactic/ball_handler_tactic.py]: I thought it should pass or shoot probabilistically, this logic is just that if there is not a good shot, then pass. This is fine, but just not what was written in the description of the tactic.

**Review by HussainGynai — CHANGES_REQUESTED** [review]
Your logic looks mostly good, although your description says it should try to pass or shoot probabilistically, but it just passes if there is no good shot. Also the tactic currently crashes as soon as the ball handler captures the ball. Also delete debug prints when done with them.

File: rj_gameplay/rj_gameplay/tactic/ball_handler_tactic.py (hunk 2 above)
Comment (kfu02) [reply rj_gameplay/rj_gameplay/tactic/ball_handler_tactic.py:107]: that's my TODO
you can delete it

**Pull request closed without merging**